File: python/alphametics/alphametics.py
from itertools import permutations
from string import digits


def solve(puzzle):
    left, right = puzzle.split("==")
    left = [s.strip() for s in left.split("+")]
    right = right.strip()
    letters = "".join(set(c for c in puzzle if c.isalpha()))

    if len(letters) > 10:
        return None

    for p in permutations(digits, len(letters)):
        trans = str.maketrans(letters, "".join(p))

        if any(len(w) > 1 and trans[ord(w[0])] == 48 for w in left + [right]):
            continue

        if sum(int(w.translate(trans)) for w in left) \
                == int(right.translate(trans)):
            return {chr(letter): digit - 48 for letter, digit in trans.items()}

    return None

File: haskell/alphametics/src/Alphametics.hs
module Alphametics
  ( solve
  ) where

import           Data.Char                      ( intToDigit
                                                , isLetter
                                                )
import           Data.List                      ( find
                                                , nub
                                                , permutations
                                                )
import qualified Data.Map                      as M
import           Data.String.Utils              ( split
                                                , strip
                                                )

solve :: String -> Maybe [(Char, Int)]
solve puzzle
  | length letters <= 10 = find f (map (zip letters) (permutations [0 .. 9]))
  | otherwise            = Nothing
 where
  letters       = nub (filter isLetter puzzle)
  [left, right] = map (map strip . split "+") (split "==" puzzle)
  f xs = all notLeadingZero (left' ++ right') && leftSum == rightSum
   where
    left'  = map (map (intToDigit . (M.fromList xs M.!))) left
    right' = map (map (intToDigit . (M.fromList xs M.!))) right
    notLeadingZero n = length n == 1 || head n /= '0'
    leftSum  = sum (map read left')
    rightSum = sum (map read right')
